pkg/api/dirs.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+	"path/filepath"
+	"strings"
+
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/manifest/jsonmanifest"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// DirUploadResponse is returned when an HTTP request to upload a directory as a tar is successful
+type DirUploadResponse struct {
+	Reference swarm.Address `json:"reference"` // manifest hash for the uploaded dir
+}
+
+// dirUploadInfo contains the data for a directory to be uploaded
+type dirUploadInfo struct {
+	toEncrypt bool
+	reader    io.ReadCloser
+}
+
+// dirUploadHandler uploads a directory supplied as a tar in an HTTP request
+func (s *server) dirUploadHandler(w http.ResponseWriter, r *http.Request) {
+	dirInfo, err := getDirHTTPInfo(r)
+	if err != nil {
+		s.Logger.Errorf("dir upload, get dir info: %v", err)
+		jsonhttp.BadRequest(w, "could not extract dir info from request")
+		return
+	}
+
+	reference, err := storeDir(r.Context(), dirInfo, s.Storer, s.Logger)
+	if err != nil {
+		s.Logger.Errorf("dir upload, store dir: %v", err)
+		jsonhttp.InternalServerError(w, "could not store dir")
+		return
+	}
+
+	jsonhttp.OK(w, DirUploadResponse{
+		Reference: reference,
+	})
+}
+
+// getDirHTTPInfo extracts data for a directory to be uploaded from an HTTP request
+func getDirHTTPInfo(r *http.Request) (*dirUploadInfo, error) {
+	if r.Body == http.NoBody {
+		return &dirUploadInfo{}, errors.New("request has no body")
+	}
+	toEncrypt := strings.ToLower(r.Header.Get(encryptHeader)) == "true"
+	return &dirUploadInfo{
+		reader:    r.Body,
+		toEncrypt: toEncrypt,
+	}, nil
+}
+
+// storeDir stores all files recursively contained in the directory given as a tar
+// it returns the hash for the uploaded manifest corresponding to the uploaded dir
+func storeDir(ctx context.Context, dirInfo *dirUploadInfo, s storage.Storer, logger logging.Logger) (swarm.Address, error) {
+	dirManifest := jsonmanifest.NewManifest()
+
+	// set up HTTP body reader
+	tarReader := tar.NewReader(dirInfo.reader)
+	defer dirInfo.reader.Close()
+
+	// iterate through the files in the supplied tar
+	for {
+		fileHeader, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return swarm.ZeroAddress, fmt.Errorf("read tar stream error: %v", err)
+		}
+
+		filePath := fileHeader.Name
+
+		// only store regular files
```

Comment:
```
// IsRegular reports whether m describes a regular file.
// That is, it tests that no mode type bits are set.
```

this means that things like directories or symbolic links will be skipped

Reply:
so this won't support empty directories. would be nice to have the support so we don't lose information when uploading. this was happening in old swarm and i think this is counter-intuitive and an undesired behavior

Reply:
sure, but can we tackle this through a subsequent PR? 

it seems like something good to iterate upon, specially since i don't really know how to implement this as of now, and don't want to drag on with this PR much further.

Reply:
i've created issue #481 for this.